config/config.go

```diff
@@ -850,92 +857,50 @@ func (c *Config) addInput(name string, table *ast.Table) error {
-		if parser, err := c.addParser(name, table); err == nil {
-			t.SetParser(parser)
-		} else {
-			missThreshold = 0
-			// Fallback to the old way of instantiating the parsers.
-			config, err := c.getParserConfig(name, table)
-			if err != nil {
-				return err
-			}
-			parser, err := c.buildParserOld(name, config)
-			if err != nil {
-				return err
-			}
-			t.SetParser(parser)
+		parser, err := c.addParser(name, table)
+		if err != nil {
+			return fmt.Errorf("adding parser failed: %w", err)
 		}
+		t.SetParser(parser)
 	}
 
 	// Keep the old interface for backward compatibility
 	if t, ok := input.(parsers.ParserInput); ok {
 		// DEPRECATED: Please switch your plugin to telegraf.ParserInput.
-		missThreshold = 1
-		if parser, err := c.addParser(name, table); err == nil {
-			t.SetParser(parser)
-		} else {
-			missThreshold = 0
-			// Fallback to the old way of instantiating the parsers.
-			config, err := c.getParserConfig(name, table)
-			if err != nil {
-				return err
-			}
-			parser, err := c.buildParserOld(name, config)
-			if err != nil {
-				return err
-			}
-			t.SetParser(parser)
+		parser, err := c.addParser(name, table)
+		if err != nil {
+			return fmt.Errorf("adding parser failed: %w", err)
 		}
+		t.SetParser(parser)
 	}
 
 	if t, ok := input.(telegraf.ParserFuncInput); ok {
-		missThreshold = 1
-		if c.probeParser(table) {
-			t.SetParserFunc(func() (telegraf.Parser, error) {
-				parser, err := c.addParser(name, table)
-				if err != nil {
-					return nil, err
-				}
-				err = parser.Init()
-				return parser, err
-			})
-		} else {
-			missThreshold = 0
-			// Fallback to the old way
-			config, err := c.getParserConfig(name, table)
+		if !c.probeParser(table) {
+			return errors.New("parser not found")
+		}
+		t.SetParserFunc(func() (telegraf.Parser, error) {
+			parser, err := c.addParser(name, table)
 			if err != nil {
-				return err
+				return nil, err
 			}
-			t.SetParserFunc(func() (telegraf.Parser, error) {
-				return c.buildParserOld(name, config)
-			})
-		}
+			err = parser.Init()
+			return parser, err
+		})
 	}
 
 	if t, ok := input.(parsers.ParserFuncInput); ok {
 		// DEPRECATED: Please switch your plugin to telegraf.ParserFuncInput.
-		missThreshold = 1
-		if c.probeParser(table) {
-			t.SetParserFunc(func() (parsers.Parser, error) {
-				parser, err := c.addParser(name, table)
-				if err != nil {
-					return nil, err
-				}
-				err = parser.Init()
-				return parser, err
-			})
-		} else {
-			missThreshold = 0
-			// Fallback to the old way
-			config, err := c.getParserConfig(name, table)
+		if !c.probeParser(table) {
+			return errors.New("parser not found")
+		}
+		t.SetParserFunc(func() (parsers.Parser, error) {
+			parser, err := c.addParser(name, table)
 			if err != nil {
-				return err
+				return nil, err
 			}
-			t.SetParserFunc(func() (parsers.Parser, error) {
-				return c.buildParserOld(name, config)
-			})
-		}
+			err = parser.Init()
+			return parser, err
+		})
 	}
```

Comment:
I think parsers.ParserFuncInput is also unused now.

Reply:
Same as above. Happy to remove it if you think we can break the `interface` contract.